### ISSUE TITLE
[no-unused-vars] False positive with template string type argument

### ISSUE THREAD

**Issue by arimah** [typescript-eslint/typescript-eslint#2495]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

Minimal repo: [arimah/typescript-eslint-no-unused-vars-issue-1](https://github.com/arimah/typescript-eslint-no-unused-vars-issue-1)

```JSON
{
  "parserOptions": {
    "sourceType": "module"
  },
  "rules": {
    "@typescript-eslint/no-unused-vars": "error"
  }
}
```

```TS
function foo<T>(x: TemplateStringsArray, ...y: T[]) {
  // (irrelevant code to get rid of unused variable warnings)
  return [x,  y];
}

interface FalsePositive { }
//        ^^^^^^^^^^^^^ @typescript-eslint/no-unused-vars - false positive

foo<FalsePositive>``;
```

**Expected Result**

No lint error: the type is used as a template string type argument.

**Actual Result**

```
6:11  error  'FalsePositive' is defined but never used  @typescript-eslint/no-unused-vars
```

**Additional Info**

False positive only occurs with `"sourceType": "module"`. With that option unset, I can't reproduce the bug, even if I introduce an explicit `export` or `import` into the code.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.0.1` |
| `@typescript-eslint/parser`        | `4.0.1` |
| `TypeScript`                       | `4.0.2` |
| `ESLint`                           | `7.8.1` |
| `node`                             | `12.16.1` |

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/2452
